clippy_utils/src/str_utils.rs

```diff
@@ -236,6 +236,30 @@ pub fn count_match_end(str1: &str, str2: &str) -> StrCount {
         })
 }
 
+/// Returns a `snake_case` version of the input
+/// ```
+/// use clippy_utils::str_utils::to_snake_case;
+/// assert_eq!(to_snake_case("AbcDef"), "abc_def");
+/// assert_eq!(to_snake_case("ABCD"), "a_b_c_d");
+/// assert_eq!(to_snake_case("AbcDD"), "abc_d_d");
+/// assert_eq!(to_snake_case("Abc1DD"), "abc1_d_d");
+/// ```
+pub fn to_snake_case(name: &str) -> String {
```

Comment:
rustc has a similar function for its nonstandard_style lint: https://doc.rust-lang.org/stable/nightly-rustc/src/rustc_lint/nonstandard_style.rs.html#237-265
it would be nice if we could somehow reuse that logic, but it might be good enough for this PR (`enum_variant` also has its own version of to_camel_case).

Reply:
I agree it would be better to use a single implementation both from `struct_fields` and for `enum_variant`. However those functions (`to_snake_case` and `to_camel_case`) are private. I think it would make sense trying to factor them out into a style utilities (similar to `clippy_utils::str_utils`) module that is accessible from clippy. 
I haven't yet contributed to rustc so I don't know if making that change for a need in clippy would be accepted. I would happily prepare a PR in rustc and clippy that unifies these implementations into a single one in rustc.
I think (from memory) there are a bunch of places in clippy that have similar ad-hoc functions related to these transformations.

Reply:
Yeah I think we could move those casing functions to some kind of util module of a crate that both clippy and rustc_lint can depend on. `rustc_ast::util::case` is an option, and if there's a better place then the reviewer over at rust-lang/rust can say it, should be trivial to move around as these are very self contained. It's probably not worth blocking this PR on syncing those two up, that could be done later on after the PR has landed.